#**Dropout Regularization** In Deep Neural Network ( credit By **Code Basics** )

This is a dataset that describes **sonar chirp** returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

In [2]:
#NumPy is a Python library used for working with arrays. It also has functions for working in domain of linear algebra, fourier transform, and matrices.
import numpy as np  
#Pandas is a Python library used for working with data sets. It has functions for analyzing, cleaning, exploring, and manipulating data.
import pandas as pd
#Matplotlib is a cross-platform, data visualization and graphical plotting library for Python and its numerical extension NumPy.
import matplotlib.pyplot as plt
#Seaborn is a Python data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics.
import seaborn as sns

In [3]:
import warnings
warnings.filterwarnings('ignore')

Read datasets

In [7]:
df = pd.read_csv("/content/drive/MyDrive/Data_Science/Datasets/sonar.all-data.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
112  0.0454  0.0472  0.0697  0.1021  0.1397  0.1493  0.1487  0.0771  0.1171   
78   0.0231  0.0351  0.0030  0.0304  0.0339  0.0860  0.1738  0.1351  0.1063   
23   0.0115  0.0150  0.0136  0.0076  0.0211  0.1058  0.1023  0.0440  0.0931   
91   0.0253  0.0808  0.0507  0.0244  0.1724  0.3823  0.3729  0.3583  0.3429   
46   0.0308  0.0339  0.0202  0.0889  0.1570  0.1750  0.0920  0.1353  0.1593   

         9   ...      51      52      53      54      55      56      57  \
112  0.1675  ...  0.0120  0.0042  0.0238  0.0129  0.0084  0.0218  0.0321   
78   0.0347  ...  0.0106  0.0097  0.0022  0.0052  0.0072  0.0056  0.0038   
23   0.0734  ...  0.0091  0.0016  0.0084  0.0064  0.0026  0.0029  0.0037   
91   0.2197  ...  0.0178  0.0073  0.0079  0.0038  0.0116  0.0033  0.0039   
46   0.2795  ...  0.0167  0.0127  0.0138  0.0090  0.0051  0.0029  0.0122   

         58      59  60  
112  0.0154  0.0053   M  
78   0.0043  0.0030   R  
23   0.0070  0.0041   R  
91   0.0081  0.0053   R  
46   0.0056  0.0020   R  

[5 rows x 61 columns]

In [8]:
df.shape

(208, 61)

In [9]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [10]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [11]:
df[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

In [12]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [13]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
35   1
59   1
14   1
34   1
193  0

In [14]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [15]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [17]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

#Using Deep Learning Model
#Model without Dropout Layer

In [18]:
import tensorflow as tf
from tensorflow import keras

In [19]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6849 - accuracy: 0.5256
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6534 - accuracy: 0.6282
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6191 - accuracy: 0.6603
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5917 - accuracy: 0.6923
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5548 - accuracy: 0.7436
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5228 - accuracy: 0.7821
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5255 - accuracy: 0.7244
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4709 - accuracy: 0.7821
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4609 - accuracy: 0.8013
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4329 - accuracy: 0.8269
Epoch 11/

In [20]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 13ms/step - loss: 0.9185 - accuracy: 0.7500


[0.9185475707054138, 0.75]

In [21]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.0960901e-07 9.8928475e-01 9.9788576e-01 2.3280681e-05 1.0000000e+00
 9.9841619e-01 3.1525016e-01 9.9999946e-01 6.1899671e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [22]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



#**Model with Dropout Layer**

In [31]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=200, batch_size=8)

Epoch 1/200
20/20 [==============================] - 1s 2ms/step - loss: 0.7055 - accuracy: 0.5513
Epoch 2/200
20/20 [==============================] - 0s 3ms/step - loss: 0.7341 - accuracy: 0.4744
Epoch 3/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6987 - accuracy: 0.5577
Epoch 4/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6960 - accuracy: 0.5192
Epoch 5/200
20/20 [==============================] - 0s 4ms/step - loss: 0.7051 - accuracy: 0.4487
Epoch 6/200
20/20 [==============================] - 0s 3ms/step - loss: 0.7063 - accuracy: 0.5256
Epoch 7/200
20/20 [==============================] - 0s 2ms/step - loss: 0.6962 - accuracy: 0.5064
Epoch 8/200
20/20 [==============================] - 0s 3ms/step - loss: 0.7030 - accuracy: 0.5385
Epoch 9/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6795 - accuracy: 0.5577
Epoch 10/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6846 - accuracy: 0.5833
Epoch 11/

In [32]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 7ms/step - loss: 0.5378 - accuracy: 0.7885


[0.5378045439720154, 0.7884615659713745]

Training Accuracy is still good but Test Accuracy Improved

In [33]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[5.8722208e-08 6.4788830e-01 8.3347666e-01 7.4130297e-04 9.9999917e-01
 9.9585867e-01 2.6134127e-01 9.9999827e-01 9.7619144e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [34]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.75      0.89      0.81        27
           1       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.81